In [311]:
# Importing required libraries
import pandas as pd

## Data Preparation

In [312]:
# Loading & showing dataset
data = pd.read_csv(filepath_or_buffer='hotel_bookings.csv')
display(data)

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
0                          0       2  ...    No Deposit    NaN     NaN   
1                          0       2  ...    No Deposit    NaN     NaN   
2                          1       1  ...    No Deposit    NaN     NaN   
3                          1       1  ...    No Deposit  304.0     NaN   
4                          2       2  ...    No Deposit  240.0     NaN   
...                      ...     ...  ...           ...    ...     ...   
119385                     5       2  ...    No Deposit  394.0     NaN   
119386                     5       3  ...    No Deposit    9.0     NaN   
119387                     5       2  ...    No Deposit    9.0     NaN   
119388                     5       2  ...    No Deposit   89.0     NaN   
119389                     7       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list customer_type     adr  \
0                         0     Transient    0.00   
1                         0     Transient    0.00   
2                         0     Transient   75.00   
3                         0     Transient   75.00   
4                         0     Transient   98.00   
...                     ...           ...     ...   
119385                    0     Transient   96.14   
119386                    0     Transient  225.43   
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
0                                 0                          0   
1                                 0                          0   
2                                 0                          0   
3             

### Context
Have you ever wondered when the best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? What if you wanted to predict whether or not a hotel was likely to receive a disproportionately high number of special requests?

This hotel booking dataset can help you explore those questions!

### Content
This [data set](https://www.kaggle.com/datasets/jessemostipak/hotel-booking-demand) contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things.

All personally identifying information has been removed from the data.

### Acknowledgements
The data is originally from the article [Hotel Booking Demand Datasets](https://www.sciencedirect.com/science/article/pii/S2352340918315191), written by Nuno Antonio, Ana Almeida, and Luis Nunes for Data in Brief, Volume 22, February 2019.

The data was downloaded and cleaned by Thomas Mock and Antoine Bichat for [#TidyTuesday during the week of February 11th, 2020](https://github.com/rfordatascience/tidytuesday/blob/master/data/2020/2020-02-11/readme.md).

### Inspiration
This data set is ideal for anyone looking to practice their exploratory data analysis (EDA) or get started in building predictive models!

If you're looking for inspiration on data visualizations, check out the [#TidyTuesday](https://github.com/rfordatascience/tidytuesday) program, a free, weekly online event that encourages participants to create and share their [code and visualizations for a given data set on Twitter](https://twitter.com/search?q=%23TidyTuesday&src=typed_query).

If you'd like to dive into predictive modeling, [Julia Silge](https://twitter.com/juliasilge) has an accessible and fantastic walk-through which highlights the [tidymodels](https://www.tidyverse.org/blog/2018/08/tidymodels-0-0-1/) R package.

### Description

| Column Name | Description |
| --- | --- |
| hotel | Hotel (H1 = Resort Hotel or H2 = City Hotel) |
| is_canceled | Value indicating if the booking was canceled (1) or not (0) |
| lead_time | Number of days that elapsed between the entering date of the booking into the PMS and the arrival date |
| arrival_date_year | Year of arrival date |
| arrival_date_month | Month of arrival date |
| arrival_date_week_number | Week number of year for arrival date |
| arrival_date_day_of_month | Day of arrival date |
| stays_in_weekend_nights | Number of weekend nights (Saturday or Sunday) the guest stayed or booked to stay at the hotel |
| stays_in_week_nights | Number of week nights (Monday to Friday) the guest stayed or booked to stay at the hotel |
| adults | Number of adults |
| children | Number of children |
| babies | Number of babies |
| meal | Type of meal booked. Categories are presented in standard hospitality meal packages: Undefined/SC – no meal package; BB – Bed & Breakfast; HB – Half board (breakfast and one other meal – usually dinner); FB – Full board (breakfast, lunch and dinner) |
| country | Country of origin. Categories are represented in the ISO 3155–3:2013 format |
| market_segment | Market segment designation. In categories, the term “TA” means “Travel Agents” and “TO” means “Tour Operators” |
| distribution_channel | Booking distribution channel. The term “TA” means “Travel Agents” and “TO” means “Tour Operators” |
| is_repeated_guest | Value indicating if the booking name was from a repeated guest (1) or not (0) |
| previous_cancellations | Number of previous bookings that were cancelled by the customer prior to the current booking |
| previous_bookings_not_canceled | Number of previous bookings not cancelled by the customer prior to the current booking |
| reserved_room_type | Code of room type reserved. Code is presented instead of designation for anonymity reasons. |
| assigned_room_type | Code for the type of room assigned to the booking. Sometimes the assigned room type differs from the reserved room type due to hotel operation reasons (e.g. overbooking) or by customer request. Code is presented instead of designation for anonymity reasons. |
| booking_changes | Number of changes/amendments made to the booking from the moment the booking was entered on the PMS until the moment of check-in or cancellation |
| deposit_type | Indication on if the customer made a deposit to guarantee the booking. This variable can assume three categories: No Deposit – no deposit was made; Non Refund – a deposit was made in the value of the total stay cost; Refundable – a deposit was made with a value under the total cost of stay. |
| agent | ID of the travel agency that made the booking |
| company | ID of the company/entity that made the booking or responsible for paying the booking. ID is presented instead of designation for anonymity reasons |
| days_in_waiting_list | Number of days the booking was in the waiting list before it was confirmed to the customer |
| customer_type | Type of booking, assuming one of four categories: Contract - when the booking has an allotment or other type of contract associated to it; Group – when the booking is associated to a group; Transient – when the booking is not part of a group or contract, and is not associated to other transient booking; Transient-party – when the booking is transient, but is associated to at least other transient booking |
| adr | Average Daily Rate as defined by dividing the sum of all lodging transactions by the total number of staying nights |
| required_car_parking_spaces | Number of car parking spaces required by the customer |
| total_of_special_requests | Number of special requests made by the customer (e.g. twin bed or high floor) |
| reservation_status | Reservation last status, assuming one of three categories: Canceled – booking was canceled by the customer; Check-Out – customer has checked in but already departed; No-Show – customer did not check-in and did inform the hotel of the reason why |
| reservation_status_date | Date at which the last status was set. This variable can be used in conjunction with the ReservationStatus to understand when was the booking canceled or when did the customer checked-out of the hotel |

---

## Data Understanding

In [313]:
# Showing more of 'Sample' column
with pd.option_context('display.max_colwidth', 225):

    # Performing initial dataset audit
    print(f'Dataset Rows, Columns: {data.shape[0]}, {data.shape[1]}')

    display(pd.DataFrame({
        'Name': data.columns.values,
        'Type': data.dtypes.values,
        'N/A (n)': data.isna().sum(),
        'N/A (%)': data.isna().mean().values * 100,
        'Unique': data.nunique().values,
        'Sample': [data[column].unique() for column in data.columns]
    }).reset_index()\
        .drop(['index'], axis=1))

Dataset Rows, Columns: 119390, 32


Name     Type  N/A (n)    N/A (%)  Unique  \
0                            hotel   object        0   0.000000       2   
1                      is_canceled    int64        0   0.000000       2   
2                        lead_time    int64        0   0.000000     479   
3                arrival_date_year    int64        0   0.000000       3   
4               arrival_date_month   object        0   0.000000      12   
5         arrival_date_week_number    int64        0   0.000000      53   
6        arrival_date_day_of_month    int64        0   0.000000      31   
7          stays_in_weekend_nights    int64        0   0.000000      17   
8             stays_in_week_nights    int64        0   0.000000      35   
9                           adults    int64        0   0.000000      14   
10                        children  float64        4   0.003350       5   
11                          babies    int64        0   0.000000       5   
12                            meal   object        0   0.000000       5   
13                         country   object      488   0.408744     177   
14                  market_segment   object        0   0.000000       8   
15            distribution_channel   object        0   0.000000       5   
16               is_repeated_guest    int64        0   0.000000       2   
17          previous_cancellations    int64        0   0.000000      15   
18  previous_bookings_not_canceled    int64        0   0.000000      73   
19              reserved_room_type   object        0   0.000000      10   
20              assigned_room_type   object        0   0.000000      12   
21                 booking_changes    int64        0   0.000000      21   
22                    deposit_type   object        0   0.000000       3   
23                           agent  float64    16340  13.686238     333   
24                         company  float64   112593  94.306893     352   
25            days_in_waiting_list    int64        0   0.000000     128   
26                   customer_type   object        0   0.000000       4   
27                             adr  float64        0   0.000000    8879   
28     required_car_parking_spaces    int64        0   0.000000       5   
29       total_of_special_requests    int64        0   0.000000       6   
30              reservation_status   object        0   0.000000       3   
31         reservation_status_date   object        0   0.000000     926   

                                                                                                                                                                                                                              Sample  
0                                                                                                                                                                                                         [Resort Hotel, City Hotel]  
1                                                                                                                                                                                                                             [0, 1]  
2   [342, 737, 7, 13, 14, 0, 9, 85, 75, 23, 35, 68, 18, 37, 12, 72, 127, 78, 48, 60, 77, 99, 118, 95, 96, 69, 45, 40, 15, 36, 43, 70, 16, 107, 47, 113, 90, 50, 93, 76, 3, 1, 10, 5, 17, 51, 71, 63, 62, 101, 2, 81, 368, 364, 32...  
3                                                                                                                                                                                                                 [2015, 2016, 2017]  
4                                                                                                                                 [July, August, September, October, November, December, January, February, March, April, May, June]  
5                        [27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11,

In [314]:
# Generating initial descriptive statistics for dataset
display(data.describe(include=['object']).T, data.describe().T)

count unique         top    freq
hotel                    119390      2  City Hotel   79330
arrival_date_month       119390     12      August   13877
meal                     119390      5          BB   92310
country                  118902    177         PRT   48590
market_segment           119390      8   Online TA   56477
distribution_channel     119390      5       TA/TO   97870
reserved_room_type       119390     10           A   85994
assigned_room_type       119390     12           A   74053
deposit_type             119390      3  No Deposit  104641
customer_type            119390      4   Transient   89613
reservation_status       119390      3   Check-Out   75166
reservation_status_date  119390    926  2015-10-21    1461

count         mean         std      min  \
is_canceled                     119390.0     0.370416    0.482918     0.00   
lead_time                       119390.0   104.011416  106.863097     0.00   
arrival_date_year               119390.0  2016.156554    0.707476  2015.00   
arrival_date_week_number        119390.0    27.165173   13.605138     1.00   
arrival_date_day_of_month       119390.0    15.798241    8.780829     1.00   
stays_in_weekend_nights         119390.0     0.927599    0.998613     0.00   
stays_in_week_nights            119390.0     2.500302    1.908286     0.00   
adults                          119390.0     1.856403    0.579261     0.00   
children                        119386.0     0.103890    0.398561     0.00   
babies                          119390.0     0.007949    0.097436     0.00   
is_repeated_guest               119390.0     0.031912    0.175767     0.00   
previous_cancellations          119390.0     0.087118    0.844336     0.00   
previous_bookings_not_canceled  119390.0     0.137097    1.497437     0.00   
booking_changes                 119390.0     0.221124    0.652306     0.00   
agent                           103050.0    86.693382  110.774548     1.00   
company                           6797.0   189.266735  131.655015     6.00   
days_in_waiting_list            119390.0     2.321149   17.594721     0.00   
adr                             119390.0   101.831122   50.535790    -6.38   
required_car_parking_spaces     119390.0     0.062518    0.245291     0.00   
total_of_special_requests       119390.0     0.571363    0.792798     0.00   

                                    25%       50%     75%     max  
is_canceled                        0.00     0.000     1.0     1.0  
lead_time                         18.00    69.000   160.0   737.0  
arrival_date_year               2016.00  2016.000  2017.0  2017.0  
arrival_date_week_number          16.00    28.000    38.0    53.0  
arrival_date_day_of_month          8.00    16.000    23.0    31.0  
stays_in_weekend_nights            0.00     1.000     2.0    19.0  
stays_in_week_nights               1.00     2.000     3.0    50.0  
adults                             2.00     2.000     2.0    55.0  
children                           0.00     0.000     0.0    10.0  
babies                             0.00     0.000     0.0    10.0  
is_repeated_guest                  0.00     0.000     0.0     1.0  
previous_cancellations             0.00     0.000     0.0    26.0  
previous_bookings_not_canceled     0.00     0.000     0.0    72.0  
booking_changes                    0.00     0.000     0.0    21.0  
agent                              9.00    14.000   229.0   535.0  
company                           62.00   179.000   270.0   543.0  
days_in_waiting_list               0.00     0.000     0.0   391.0  
adr                               69.29    94.575   126.0  5400.0  
required_car_parking_spaces        0.00     0.000     0.0     8.0  
total_of_special_requests          0.00     0.000     1.0     5.0

### Data Insights

| Column Name | Insights |
| --- | --- |
| hotel | - |
| is_canceled | Invalid data type: `bool` instead of `int` based on data description |
| lead_time | - |
|  | Outlier: `737.0` is way above the mean, standard deviation, and 75% value of the data |
| arrival_date_year | May be merged using datetime format with `arrival_date_month` and `arrival_date_day_of_month` into `arrival_date` |
| arrival_date_month | May be merged using datetime format with `arrival_date_year` and `arrival_date_day_of_month` into `arrival_date` |
| arrival_date_week_number | May be dropped: redundant with `arrival_date` |
| arrival_date_day_of_month | May be merged using datetime format with `arrival_date_year` and `arrival_date_month` into `arrival_date` |
| stays_in_weekend_nights | Outlier: `53.0` is way above the mean, standard deviation, and 75% value of the data |
| stays_in_week_nights | Outlier: `31.0` is way above the mean, standard deviation, and 75% value of the data |
| adults | Outlier: `55.0` is way above the mean, standard deviation, and 75% value of the data |
| children | Invalid data type: `int` instead of `float` based on data description |
|  | Missing value: `nan` |
|  | Outlier: `10.0` is way above the mean, standard deviation, and 75% value of the data |
| babies | Outlier: `10.0` is way above the mean, standard deviation, and 75% value of the data |
| meal | Invalid value: `Undefined` —may be recategorized as `SC` based on data description |
| country | Missing value: `nan` |
| market_segment | - |
| distribution_channel | - |
| is_repeated_guest | Invalid data type: `bool` instead of `int` based on data description |
| previous_cancellations | Outlier: `26.0` is way above the mean, standard deviation, and 75% value of the data |
| previous_bookings_not_canceled | Outlier: `72.0` is way above the mean, standard deviation, and 75% value of the data |
| reserved_room_type | - |
| assigned_room_type | - |
| booking_changes | - |
| deposit_type | - |
| agent | Invalid data type: `object` instead of `float` based on data description |
|  | Missing value: `nan` |
| company | Invalid data type: `object` instead of `float` based on data description |
|  | Missing value: `nan` |
| days_in_waiting_list | Outlier: `391.0` is way above the mean, standard deviation, and 75% value of the data |
| customer_type | - |
| adr | Outlier: `-6.38` and `5400.0` is way beyond the mean, standard deviation, and 25% or 75% value of the data, respectively |
| required_car_parking_spaces | Outlier: `8.0` is way above the mean, standard deviation, and 75% value of the data|
| total_of_special_requests | Outlier: `5.0` is way above the mean, standard deviation, and 75% value of the data|
| reservation_status | - |
| reservation_status_date | Invalid data type: `datetime` instead of `object` based on data description |

### Notes

* `PMS` = Property Management System
* `country` nan > UNK? 
* `market_segment`?
* `GDS` = Global Distribution System
* potential `reserved_room_type` and `assigned_room_type` redundant with `booking_changes`?
* `agent` and `company` may share similar values per entry/index?
* transient business consists of corporate and leisure guests looking for short and last-minute hotel stays.
* If a hotel has $50,000 in room revenue and 500 rooms sold, the ADR would be $100 ($50,000/500).
* when the best time of year to book a hotel room is = ?
* the optimal length of stay in order to get the best daily rate = ? 
* whether or not a hotel was likely to receive a disproportionately high number of special requests = `total_of_special_requests`?

---

# Data Cleaning

In [315]:
# Making a copy of dataset
df = data.copy()

* To avoid chained indexing

## Duplicated Data

In [316]:
# Detecting any duplicated data
df[df.duplicated()]

hotel  is_canceled  lead_time  arrival_date_year  \
5       Resort Hotel            0         14               2015   
22      Resort Hotel            0         72               2015   
43      Resort Hotel            0         70               2015   
138     Resort Hotel            1          5               2015   
200     Resort Hotel            0          0               2015   
...              ...          ...        ...                ...   
119349    City Hotel            0        186               2017   
119352    City Hotel            0         63               2017   
119353    City Hotel            0         63               2017   
119354    City Hotel            0         63               2017   
119373    City Hotel            0        175               2017   

       arrival_date_month  arrival_date_week_number  \
5                    July                        27   
22                   July                        27   
43                   July                        27   
138                  July                        28   
200                  July                        28   
...                   ...                       ...   
119349             August                        35   
119352             August                        35   
119353             August                        35   
119354             August                        35   
119373             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
5                               1                        0   
22                              1                        2   
43                              2                        2   
138                             5                        1   
200                             7                        0   
...                           ...                      ...   
119349                         31                        0   
119352                         31                        0   
119353                         31                        0   
119354                         31                        0   
119373                         31                        1   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
5                          2       2  ...    No Deposit  240.0     NaN   
22                         4       2  ...    No Deposit  250.0     NaN   
43                         3       2  ...    No Deposit  250.0     NaN   
138                        0       2  ...    No Deposit  240.0     NaN   
200                        1       1  ...    No Deposit  240.0     NaN   
...                      ...     ...  ...           ...    ...     ...   
119349                     3       2  ...    No Deposit    9.0     NaN   
119352                     3       3  ...    No Deposit    9.0     NaN   
119353                     3       3  ...    No Deposit    9.0     NaN   
119354                     3       3  ...    No Deposit    9.0     NaN   
119373                     3       1  ...    No Deposit   42.0     NaN   

       days_in_waiting_list    customer_type     adr  \
5                         0        Transient   98.00   
22                        0        Transient   84.67   
43                        0        Transient  137.00   
138                       0        Transient   97.00   
200                       0        Transient  109.80   
...                     ...              ...     ...   
119349                    0        Transient  126.00   
119352                    0  Transient-Party  195.33   
119353                    0  Transient-Party  195.33   
119354                    0  Transient-Party  195.33   
119373                    0        Transient   82.35   

        required_car_parking_spaces  total_of_special_requests  \
5                                 0                          1   
22                                0                          1   
43                                0         

In [317]:
# Dropping duplicated data
df = df.drop_duplicates(keep='first') # Keeping first instances of duplicated data

# Verifying process
df[df.duplicated()]

Empty DataFrame
Columns: [hotel, is_canceled, lead_time, arrival_date_year, arrival_date_month, arrival_date_week_number, arrival_date_day_of_month, stays_in_weekend_nights, stays_in_week_nights, adults, children, babies, meal, country, market_segment, distribution_channel, is_repeated_guest, previous_cancellations, previous_bookings_not_canceled, reserved_room_type, assigned_room_type, booking_changes, deposit_type, agent, company, days_in_waiting_list, customer_type, adr, required_car_parking_spaces, total_of_special_requests, reservation_status, reservation_status_date]
Index: []

[0 rows x 32 columns]

## Column Merge

In [ ]:
# Getting integer location of 'arrival_date_year'
arrival_date_year_loc = df.columns.get_loc(key='arrival_date_year')

# Casting 'arrival_date_year' and 'arrival_date_day_of_month' to string data type
df = df.astype(dtype={'arrival_date_year': str,
                      'arrival_date_day_of_month': str})

# Inserting 'arrival_date' at 'arrival_date_year' location
df.insert(loc=arrival_date_year_loc, column='arrival_date', value=df.arrival_date_day_of_month + ' ' + df.arrival_date_month + ' ' + df.arrival_date_year)

# Dropping unused columns
df.drop(columns=['arrival_date_year', 'arrival_date_month', 'arrival_date_week_number', 'arrival_date_day_of_month'], inplace=True)

# Verifying process
df.head()

hotel  is_canceled  lead_time arrival_date  stays_in_weekend_nights  \
0  Resort Hotel            0        342  1 July 2015                        0   
1  Resort Hotel            0        737  1 July 2015                        0   
2  Resort Hotel            0          7  1 July 2015                        0   
3  Resort Hotel            0         13  1 July 2015                        0   
4  Resort Hotel            0         14  1 July 2015                        0   

   stays_in_week_nights  adults  children  babies meal  ... deposit_type  \
0                     0       2       0.0       0   BB  ...   No Deposit   
1                     0       2       0.0       0   BB  ...   No Deposit   
2                     1       1       0.0       0   BB  ...   No Deposit   
3                     1       1       0.0       0   BB  ...   No Deposit   
4                     2       2       0.0       0   BB  ...   No Deposit   

   agent company  days_in_waiting_list  customer_type   adr  \
0    NaN     NaN                     0      Transient   0.0   
1    NaN     NaN                     0      Transient   0.0   
2    NaN     NaN                     0      Transient  75.0   
3  304.0     NaN                     0      Transient  75.0   
4  240.0     NaN                     0      Transient  98.0   

  required_car_parking_spaces total_of_special_requests  reservation_status  \
0                           0                         0           Check-Out   
1                           0                         0           Check-Out   
2                           0                         0           Check-Out   
3                           0                         0           Check-Out   
4                           0                         1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 29 columns]

## Invalid Data

In [ ]:
# Casting columns to their appropriate data type
df = df.astype(dtype={'is_canceled': 'bool',
                      'children': 'Int64',
                      'is_repeated_guest': 'bool',
                      'agent': 'Int64',
                      'company': 'Int64'})\
       .astype(dtype={'agent': 'str',
                      'company': 'str'})

# Converting columns to datetime 
datetime_columns = ['arrival_date', 'reservation_status_date']
for column in datetime_columns:
    df[column] = pd.to_datetime(arg=df[column])

# Replacing values in 'meal'
df.replace(to_replace={'meal': {'Undefined': 'SC'}}, inplace=True)